python/pyproject.toml

```diff
@@ -0,0 +1,27 @@
+[build-system]
+requires = ["setuptools>=42", "cython", "cmake"]
+build-backend = "setuptools.build_meta"
+
+[tool.cibuildwheel]
+build = 'cp3{9,10}-manylinux_x86_64'
+manylinux-x86_64-image = 'manylinux_2_28'
+before-all = [
+  'dnf install -y epel-release',
+  'dnf module enable -y mariadb-devel',
+  'dnf install -y $(grep "^[^#]" dependencies.txt)',
+  'echo "need newer asio-devel version"',
+  'curl -OL https://download.sourceforge.net/asio/asio-1.21.0.tar.bz2',
+  'tar xf asio-1.21.0.tar.bz2',
+  'cd asio-1.21.0',
+  './configure',
+  'make',
+  'make install',
+]
```

Comment:
There are many things I find unsettling here.

1. Python level file, which is supposed to be built in user space, contains system commands that alter my OS.

2. Encoding bash scripts in TOML like that. The thought that I may need to copy/paste these commands for debugging. TOML supports multiline strings, so why not to use it here to be on par with `CIBW_BEFORE_ALL`?

3. `dependencies.txt` is in parent dir, so `dnf install` won't install anything and should fail. It if doesn't then that's a problem.

4.  After `cd asio-1.21.0` there is no `cd` back, so it is not clear in which dir `cibuidwheel` should start the process.

5. The `asio` build process is only valid for before all on `manylinux_2_28`, because AlmaLinux 8 ships with outdated dependency.

So in the context of CI job having such hack is fine, but embedding it in Python project doesn't look good to me. So I think about moving shell parts into standalone build scripts, but then there is already CMake that is for doing builds. Oh my.. :D

Reply:
1: No it does not. This only runs inside the docker container, even locally. But you your point, it should be prefixed by linux just in case. macOS and Windows would run on your OS. And these lines wouldn't be valid anyway.

2: Multiline strings would require the &&'ing together. But you could use that if you prefer. If you are debugging, you are probably running line by line anyway, or you are running cibuldwheel locally, in which there's no copy/paste.

3: This is run from the parent dir. It's being run by cibuildwheel.

4: before-build is not tied to the per-build runs. It _can't_ be, it runs once while the per-build runs run on each time. They are separate processes (with seperate Pythons, etc). So no need to clean up cd's.

5: You only support manylinux_2_28 right?

This is cibulidwheel config. Just like you had before. Only now it's shorter and can be run locally.

Reply:
I built the wheels locally to make sure this worked. That's how I was able to do it in a few minutes instead of 7 days of back & forth with the CI. There were a few oddities I didn't expect (like Cython being required twice).

Reply:
Multi-line strings should obviously contain newline characters, which, when copy-pasted into a shell, are equivalent to `;`, which is likewise equivalent to `&&` except that one failing command will not prevent newline / `;` from running the follow-up commands.